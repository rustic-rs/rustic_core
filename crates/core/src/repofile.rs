use std::ops::Deref;

use serde::{de::DeserializeOwned, Serialize};

pub(crate) mod configfile;
pub(crate) mod indexfile;
pub(crate) mod keyfile;
pub(crate) mod packfile;
pub(crate) mod snapshotfile;

/// Marker trait for repository files which are stored as encrypted JSON
pub trait RepoFile: Serialize + DeserializeOwned + Sized + Send + Sync + 'static {
    /// The [`FileType`] associated with the repository file
    const TYPE: FileType;
    /// The Id type associated with the repository file
    type Id: From<Id> + Send;
}

/// Marker trait for Ids which identify repository files
pub trait RepoId: Deref<Target = Id> + From<Id> + Sized + Send + Sync + 'static {
    /// The [`FileType`] associated with Id type
    const TYPE: FileType;
}

#[macro_export]
/// Generate newtypes for `Id`s identifying Repository files
macro_rules! impl_repoid {
    ($a:ident, $b: expr) => {
        $crate::define_new_id_struct!($a, concat!("repository file of type", stringify!($b)));
        impl $crate::repofile::RepoId for $a {
            const TYPE: FileType = $b;
        }
    };
}

#[macro_export]
/// Generate newtypes for `Id`s identifying Repository files implementing `RepoFile`
macro_rules! impl_repofile {
    ($a:ident, $b: expr, $c: ty) => {
        $crate::impl_repoid!($a, $b);
        impl RepoFile for $c {
            const TYPE: FileType = $b;
            type Id = $a;
        }
    };
}

// Part of public API
use crate::Id;

pub use {
    crate::{
        backend::{
            node::{Metadata, Node, NodeType},
            FileType, ALL_FILE_TYPES,
        },
        blob::{tree::Tree, BlobType, ALL_BLOB_TYPES},
    },
    configfile::ConfigFile,
    indexfile::{IndexBlob, IndexFile, IndexId, IndexPack},
    keyfile::{KeyFile, KeyId},
    packfile::{HeaderEntry, PackHeader, PackHeaderLength, PackHeaderRef, PackId},
    snapshotfile::{DeleteOption, PathList, SnapshotFile, SnapshotId, SnapshotSummary, StringList},
};
