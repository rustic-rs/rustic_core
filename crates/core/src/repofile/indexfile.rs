use std::{cmp::Ordering, num::NonZeroU32};

use chrono::{DateTime, Local};
use serde_derive::{Deserialize, Serialize};
use serde_with::skip_serializing_none;

use crate::{
    backend::FileType,
    blob::{BlobId, BlobType},
    impl_repoid,
    repofile::{packfile::PackHeaderRef, RepoFile},
};

use super::packfile::PackId;

impl_repoid!(IndexId, FileType::Index);

/// Index files describe index information about multiple `pack` files.
///
/// They are usually stored in the repository under `/index/<ID>`
#[skip_serializing_none]
#[derive(Serialize, Deserialize, Debug, Default)]
pub struct IndexFile {
    /// which other index files are superseded by this (not actively used)
    pub supersedes: Option<Vec<IndexId>>,
    /// Index information about used packs
    pub packs: Vec<IndexPack>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    /// Index information about unused packs which are already marked for deletion
    pub packs_to_delete: Vec<IndexPack>,
}

impl RepoFile for IndexFile {
    /// The [`FileType`] associated with the [`IndexFile`]
    const TYPE: FileType = FileType::Index;
    type Id = IndexId;
}

impl IndexFile {
    /// Add a new pack to the index file
    ///
    /// # Arguments
    ///
    /// * `p` - The pack to add
    /// * `delete` - If the pack should be marked for deletion
    pub(crate) fn add(&mut self, p: IndexPack, delete: bool) {
        if delete {
            self.packs_to_delete.push(p);
        } else {
            self.packs.push(p);
        }
    }

    pub(crate) fn all_packs(self) -> impl Iterator<Item = (IndexPack, bool)> {
        self.packs
            .into_iter()
            .map(|pack| (pack, false))
            .chain(self.packs_to_delete.into_iter().map(|pack| (pack, true)))
    }
}

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Default, Debug, Clone)]
/// Index information about a `pack`
pub struct IndexPack {
    /// pack Id
    pub id: PackId,
    /// Index information about contained blobs
    pub blobs: Vec<IndexBlob>,
    /// The pack creation time or time when the pack was marked for deletion
    pub time: Option<DateTime<Local>>,
    /// The pack size
    pub size: Option<u32>,
}

impl IndexPack {
    /// Add a new blob to the pack
    ///
    /// # Arguments
    ///
    /// * `id` - The blob id
    /// * `tpe` - The blob type
    /// * `offset` - The blob offset within the pack
    /// * `length` - The blob length within the pack
    /// * `uncompressed_length` - The blob uncompressed length within the pack
    pub(crate) fn add(
        &mut self,
        id: BlobId,
        tpe: BlobType,
        offset: u32,
        length: u32,
        uncompressed_length: Option<NonZeroU32>,
    ) {
        self.blobs.push(IndexBlob {
            id,
            tpe,
            offset,
            length,
            uncompressed_length,
        });
    }

    /// Calculate the pack size from the contained blobs
    #[must_use]
    pub fn pack_size(&self) -> u32 {
        self.size
            .unwrap_or_else(|| PackHeaderRef::from_index_pack(self).pack_size())
    }

    /// Returns the blob type of the pack.
    ///
    /// # Note
    ///
    /// Only packs with identical blob types are allowed.
    #[must_use]
    pub fn blob_type(&self) -> BlobType {
        // TODO: This is a hack to support packs without blobs (e.g. when deleting unreferenced files)
        if self.blobs.is_empty() {
            BlobType::Data
        } else {
            self.blobs[0].tpe
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, Copy)]
/// Index information about a `blob`
pub struct IndexBlob {
    /// Blob Id
    pub id: BlobId,
    #[serde(rename = "type")]
    /// Type of the blob
    pub tpe: BlobType,
    /// Offset of the blob within the `pack` file
    pub offset: u32,
    /// Length of the blob as stored within the `pack` file
    pub length: u32,
    /// Data length of the blob. This is only set if the blob is compressed.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub uncompressed_length: Option<NonZeroU32>,
}

impl PartialOrd<Self> for IndexBlob {
    /// Compare two blobs by their offset
    ///
    /// # Arguments
    ///
    /// * `other` - The other blob to compare to
    ///
    /// # Returns
    ///
    /// The ordering of the two blobs
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for IndexBlob {
    /// Compare two blobs by their offset
    ///
    /// # Arguments
    ///
    /// * `other` - The other blob to compare to
    ///
    /// # Returns
    ///
    /// The ordering of the two blobs
    fn cmp(&self, other: &Self) -> Ordering {
        self.offset.cmp(&other.offset)
    }
}
