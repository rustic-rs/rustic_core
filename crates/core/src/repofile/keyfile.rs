use chrono::{DateTime, Local};
use rand::{thread_rng, RngCore};
use scrypt::Params;
use serde_derive::{Deserialize, Serialize};
use serde_with::{base64::Base64, serde_as, skip_serializing_none};

use crate::{
    backend::{FileType, ReadBackend},
    crypto::{aespoly1305::Key, CryptoKey},
    error::{ErrorKind, RusticError, RusticResult},
    impl_repoid,
};

/// [`KeyFileErrorKind`] describes the errors that can be returned for `KeyFile`s
#[derive(thiserror::Error, Debug, displaydoc::Display)]
#[non_exhaustive]
pub enum KeyFileErrorKind {
    /// conversion from `{from}` to `{to}` failed for `{x}` : `{source}`
    ConversionFailed {
        from: &'static str,
        to: &'static str,
        x: u32,
        source: std::num::TryFromIntError,
    },
}

pub(crate) type KeyFileResult<T> = Result<T, KeyFileErrorKind>;

pub(super) mod constants {
    /// Returns the number of bits of the given type.
    pub(super) const fn num_bits<T>() -> usize {
        // Needed for MSRV 1.76
        #![allow(unused_qualifications)]
        std::mem::size_of::<T>() * 8
    }
}

impl_repoid!(KeyId, FileType::Key);

/// Key files describe information about repository access keys.
///
/// They are usually stored in the repository under `/keys/<ID>`
#[serde_as]
#[skip_serializing_none]
#[derive(Serialize, Deserialize, Debug)]
pub struct KeyFile {
    /// Hostname where the key was created
    hostname: Option<String>,

    /// User which created the key
    username: Option<String>,

    /// Creation time of the key
    created: Option<DateTime<Local>>,

    /// The used key derivation function (currently only `scrypt`)
    kdf: String,

    /// Parameter N for `scrypt`
    #[serde(rename = "N")]
    n: u32,

    /// Parameter r for `scrypt`
    r: u32,

    /// Parameter p for `scrypt`
    p: u32,

    /// The key data encrypted by `scrypt`
    #[serde_as(as = "Base64")]
    data: Vec<u8>,

    /// The salt used with `scrypt`
    #[serde_as(as = "Base64")]
    salt: Vec<u8>,
}

impl KeyFile {
    /// Generate a Key using the key derivation function from [`KeyFile`] and a given password
    ///
    /// # Arguments
    ///
    /// * `passwd` - The password to use for the key derivation function
    ///
    /// # Errors
    ///
    /// * If the parameters of the key derivation function are invalid
    /// * If the output length of the key derivation function is invalid
    ///
    /// # Returns
    ///
    /// The generated key
    pub fn kdf_key(&self, passwd: &impl AsRef<[u8]>) -> RusticResult<Key> {
        let params = Params::new(
            log_2(self.n).map_err(|err| {
                RusticError::with_source(
                    ErrorKind::Internal,
                    "Calculating log2 failed. Please check the key file and password.",
                    err,
                )
            })?,
            self.r,
            self.p,
            Params::RECOMMENDED_LEN,
        )
        .map_err(|err| {
            RusticError::with_source(
                ErrorKind::Key,
                "Invalid scrypt parameters. Please check the key file and password.",
                err,
            )
        })?;

        let mut key = [0; 64];
        scrypt::scrypt(passwd.as_ref(), &self.salt, &params, &mut key).map_err(|err| {
            RusticError::with_source(
                ErrorKind::Key,
                "Output length invalid. Please check the key file and password.",
                err,
            )
        })?;

        Ok(Key::from_slice(&key))
    }

    /// Extract a key from the data of the [`KeyFile`] using the given key.
    /// The key usually should be the key generated by [`kdf_key()`](Self::kdf_key)
    ///
    /// # Arguments
    ///
    /// * `key` - The key to use for decryption
    ///
    /// # Errors
    ///
    /// * If the data could not be deserialized
    ///
    /// # Returns
    ///
    /// The extracted key
    pub fn key_from_data(&self, key: &Key) -> RusticResult<Key> {
        let dec_data = key.decrypt_data(&self.data)?;

        let key = serde_json::from_slice::<MasterKey>(&dec_data)
            .map_err(|err| {
                RusticError::with_source(
                    ErrorKind::Key,
                    "Deserializing master key from slice failed. Please check the key file.",
                    err,
                )
            })?
            .key();

        Ok(key)
    }

    /// Extract a key from the data of the [`KeyFile`] using the key
    /// from the derivation function in combination with the given password.
    ///
    /// # Arguments
    ///
    /// * `passwd` - The password to use for the key derivation function
    ///
    /// # Errors
    ///
    /// * If the parameters of the key derivation function are invalid
    ///
    /// # Returns
    ///
    /// The extracted key
    pub fn key_from_password(&self, passwd: &impl AsRef<[u8]>) -> RusticResult<Key> {
        self.key_from_data(&self.kdf_key(passwd)?)
    }

    /// Generate a new [`KeyFile`] from a given key and password.
    ///
    /// # Arguments
    ///
    /// * `key` - The key to use for encryption
    /// * `passwd` - The password to use for the key derivation function
    /// * `hostname` - The hostname to use for the [`KeyFile`]
    /// * `username` - The username to use for the [`KeyFile`]
    /// * `with_created` - Whether to set the creation time of the [`KeyFile`] to the current time
    ///
    /// # Errors
    ///
    /// * If the output length of the key derivation function is invalid
    /// * If the [`KeyFile`] could not be serialized
    ///
    /// # Returns
    ///
    /// The generated [`KeyFile`]
    pub fn generate(
        key: Key,
        passwd: &impl AsRef<[u8]>,
        hostname: Option<String>,
        username: Option<String>,
        with_created: bool,
    ) -> RusticResult<Self> {
        let masterkey = MasterKey::from_key(key);
        let params = Params::recommended();
        let mut salt = vec![0; 64];
        thread_rng().fill_bytes(&mut salt);

        let mut key = [0; 64];
        scrypt::scrypt(passwd.as_ref(), &salt, &params, &mut key).map_err(|err| {
            RusticError::with_source(
                ErrorKind::Key,
                "Output length invalid. Please check the key file and password.",
                err,
            )
        })?;

        let key = Key::from_slice(&key);

        let json_byte_vec = serde_json::to_vec(&masterkey).map_err(|err| {
            RusticError::with_source(
                ErrorKind::Key,
                "Could not serialize as JSON byte vector.",
                err,
            )
            .ask_report()
        })?;

        let data = key.encrypt_data(&json_byte_vec)?;

        Ok(Self {
            hostname,
            username,
            kdf: "scrypt".to_string(),
            n: 2_u32.pow(u32::from(params.log_n())),
            r: params.r(),
            p: params.p(),
            created: with_created.then(Local::now),
            data,
            salt,
        })
    }

    /// Get a [`KeyFile`] from the backend
    ///
    /// # Arguments
    ///
    /// * `be` - The backend to use
    /// * `id` - The id of the [`KeyFile`]
    ///
    /// # Errors
    ///
    /// * If the [`KeyFile`] could not be deserialized/read from the backend
    ///
    /// # Returns
    ///
    /// The [`KeyFile`] read from the backend
    fn from_backend<B: ReadBackend>(be: &B, id: &KeyId) -> RusticResult<Self> {
        let data = be.read_full(FileType::Key, id)?;

        serde_json::from_slice(&data).map_err(|err| {
            RusticError::with_source(
                ErrorKind::Key,
                "Couldn't deserialize the data for key `{key_id}`.",
                err,
            )
            .attach_context("key_id", id.to_string())
        })
    }
}

/// Calculate the logarithm to base 2 of the given number
///
/// # Arguments
///
/// * `x` - The number to calculate the logarithm to base 2 of
///
/// # Errors
///
/// * If the conversion from `u32` to `u8` failed
///
/// # Returns
///
/// The logarithm to base 2 of the given number
fn log_2(x: u32) -> KeyFileResult<u8> {
    assert!(x > 0);
    Ok(u8::try_from(constants::num_bits::<u32>()).map_err(|err| {
        KeyFileErrorKind::ConversionFailed {
            from: "usize",
            to: "u8",
            x,
            source: err,
        }
    })? - u8::try_from(x.leading_zeros()).map_err(|err| KeyFileErrorKind::ConversionFailed {
        from: "u32",
        to: "u8",
        x,
        source: err,
    })? - 1)
}

/// The mac of a [`Key`]
///
/// This is used to verify the integrity of the key
#[serde_as]
#[derive(Serialize, Deserialize, Debug)]
pub(crate) struct Mac {
    /// The key used for the mac
    #[serde_as(as = "Base64")]
    k: Vec<u8>,

    /// The random value used for the mac
    #[serde_as(as = "Base64")]
    r: Vec<u8>,
}

/// The master key of a [`Key`]
///
/// This is used to encrypt the key
#[serde_as]
#[derive(Serialize, Deserialize, Debug)]
pub(crate) struct MasterKey {
    /// The mac of the key
    mac: Mac,
    /// The encrypted key
    #[serde_as(as = "Base64")]
    encrypt: Vec<u8>,
}

impl MasterKey {
    /// Create a [`MasterKey`] from a [`Key`]
    ///
    /// # Arguments
    ///
    /// * `key` - The key to create the [`MasterKey`] from
    ///
    /// # Returns
    ///
    /// The created [`MasterKey`]
    fn from_key(key: Key) -> Self {
        let (encrypt, k, r) = key.to_keys();
        Self {
            encrypt,
            mac: Mac { k, r },
        }
    }

    /// Get the [`Key`] from the [`MasterKey`]
    fn key(&self) -> Key {
        Key::from_keys(&self.encrypt, &self.mac.k, &self.mac.r)
    }
}

/// Get a [`KeyFile`] from the backend
///
/// # Arguments
///
/// * `be` - The backend to use
/// * `id` - The id of the [`KeyFile`]
/// * `passwd` - The password to use
///
/// # Errors
///
// TODO!: Add errors!
pub(crate) fn key_from_backend<B: ReadBackend>(
    be: &B,
    id: &KeyId,
    passwd: &impl AsRef<[u8]>,
) -> RusticResult<Key> {
    KeyFile::from_backend(be, id)?.key_from_password(passwd)
}

/// Find a [`KeyFile`] in the backend that fits to the given password and return the contained key.
/// If a key hint is given, only this key is tested.
/// This is recommended for a large number of keys.
///
/// # Arguments
///
/// * `be` - The backend to use
/// * `passwd` - The password to use
/// * `hint` - The key hint to use
///
/// # Errors
///
/// * If no suitable key was found
///
/// # Returns
///
/// The found key
pub(crate) fn find_key_in_backend<B: ReadBackend>(
    be: &B,
    passwd: &impl AsRef<[u8]>,
    hint: Option<&KeyId>,
) -> RusticResult<Key> {
    if let Some(id) = hint {
        key_from_backend(be, id, passwd)
    } else {
        for id in be.list(FileType::Key)? {
            match key_from_backend(be, &id.into(), passwd) {
                Ok(key) => return Ok(key),
                Err(err) if err.is_code("C001") => continue,
                err => return err,
            }
        }

        Err(RusticError::new(
            ErrorKind::Password,
            "The password that has been entered, seems to be incorrect. No suitable key found for the given password. Please check your password and try again.",
        ).attach_error_code("C002"))
    }
}
