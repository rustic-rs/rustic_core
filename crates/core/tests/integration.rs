//! Integration tests for the core library
//!
//! # How to update snapshots
//!
//! The CI pipeline is configured to run the tests with the `INSTA_UPDATE` environment variable set to `new`.
//! This means, it uploads the failed tests snapshots to the artifacts and you can download them and use them to update the snapshots.
//!
//! To update the snapshots, you download the artifacts and copy the files to the `tests/snapshots` directory.
//! Then you run `cargo insta review` to review the changes and accept them.
//!
//! # Redactions
//!
//! We use the `insta` crate to compare the actual output of the tests with the expected output.
//! Some data in the output changes every test run, we use insta's redactions to replace the actual values with placeholders.
//! We define the redactions inside `Settings` in the fixtures and bind them to the test functions. You can read more about
//! it [here](https://docs.rs/insta/latest/insta/struct.Settings.html).
//!
//! # Fixtures and Dependency Injection
//!
//! We use the `rstest` crate to define fixtures and dependency injection.
//! This allows us to define a set of fixtures that are used in multiple tests.
//! The fixtures are defined as functions with the `#[fixture]` attribute.
//! The tests that use the fixtures are defined as functions with the `#[rstest]` attribute.
//! The fixtures are passed as arguments to the test functions.

use anyhow::Result;
use flate2::read::GzDecoder;
use globset::Glob;
use insta::internals::{Content, ContentPath};
use insta::{assert_ron_snapshot, Settings};
use pretty_assertions::assert_eq;
use rstest::fixture;
use rstest::rstest;
use rustic_core::repofile::{Metadata, Node};
use rustic_core::{
    repofile::SnapshotFile, BackupOptions, ConfigOptions, KeyOptions, LsOptions, NoProgressBars,
    OpenStatus, PathList, Repository, RepositoryBackends, RepositoryOptions,
};
use rustic_core::{FindMatches, FindNode, RusticResult};
use serde::Serialize;

use rustic_testing::backend::in_memory_backend::InMemoryBackend;

use std::ffi::OsStr;
use std::{
    env,
    fs::File,
    path::{Path, PathBuf},
    str::FromStr,
    sync::Arc,
};
// uncomment for logging output
// use simplelog::{Config, SimpleLogger};
use tar::Archive;
use tempfile::{tempdir, TempDir};

type RepoOpen = Repository<NoProgressBars, OpenStatus>;

#[derive(Debug)]
struct TestSource(TempDir);

impl TestSource {
    fn new(tmp: TempDir) -> Self {
        Self(tmp)
    }

    fn path_list(&self) -> PathList {
        PathList::from_iter(Some(self.0.path().to_path_buf()))
    }
}

#[fixture]
fn set_up_repo() -> Result<RepoOpen> {
    let be = InMemoryBackend::new();
    let be = RepositoryBackends::new(Arc::new(be), None);
    let options = RepositoryOptions::default().password("test");
    let repo = Repository::new(&options, &be)?;
    let key_opts = KeyOptions::default();
    let config_opts = &ConfigOptions::default();
    let repo = repo.init(&key_opts, config_opts)?;
    Ok(repo)
}

// helper func to redact options, but still keep information about some/none
#[allow(clippy::needless_pass_by_value)] // we need exactly that function signature
fn handle_option(val: Content, _: ContentPath<'_>) -> String {
    if val.is_nil() {
        "[none]".to_string()
    } else {
        "[some]".to_string()
    }
}

#[fixture]
fn insta_summary_redaction() -> Settings {
    let mut settings = insta::Settings::clone_current();

    settings.add_redaction(".tree", "[tree_id]");
    settings.add_dynamic_redaction(".program_version", |val, _| {
        val.resolve_inner()
            .as_str()
            .map_or("[program_version]".to_string(), |v| {
                v.replace(env!("CARGO_PKG_VERSION"), "[rustic_core_version]")
            })
    });
    settings.add_redaction(".time", "[time]");
    settings.add_dynamic_redaction(".parent", handle_option);
    settings.add_redaction(".tags", "[tags]");
    settings.add_redaction(".id", "[id]");
    settings.add_redaction(".summary.backup_start", "[backup_start]");
    settings.add_redaction(".summary.backup_end", "[backup_end]");
    settings.add_redaction(".summary.backup_duration", "[backup_duration]");
    settings.add_redaction(".summary.total_duration", "[total_duration]");
    settings.add_redaction(".summary.data_added", "[data_added]");
    settings.add_redaction(".summary.data_added_packed", "[data_added_packed]");
    settings.add_redaction(
        ".summary.total_dirsize_processed",
        "[total_dirsize_processed]",
    );
    settings.add_redaction(
        ".summary.data_added_trees_packed",
        "[data_added_trees_packed]",
    );
    settings.add_redaction(".summary.data_added_trees", "[data_added_trees]");

    settings
}

#[fixture]
fn insta_tree_redaction() -> Settings {
    let mut settings = insta::Settings::clone_current();

    settings.add_redaction(".nodes[].inode", "[inode]");
    settings.add_redaction(".nodes[].device_id", "[device_id]");
    settings.add_redaction(".nodes[].uid", "[uid]");
    settings.add_redaction(".nodes[].user", "[user]");
    settings.add_redaction(".nodes[].gid", "[gid]");
    settings.add_redaction(".nodes[].group", "[group]");
    settings.add_dynamic_redaction(".nodes[].mode", handle_option);
    settings.add_dynamic_redaction(".nodes[].mtime", handle_option);
    settings.add_dynamic_redaction(".nodes[].atime", handle_option);
    settings.add_dynamic_redaction(".nodes[].ctime", handle_option);

    settings
}

#[fixture]
fn tar_gz_testdata() -> Result<TestSource> {
    let dir = tempdir()?;
    let path = Path::new("tests/fixtures/backup-data.tar.gz").canonicalize()?;
    let tar_gz = File::open(path)?;
    let tar = GzDecoder::new(tar_gz);
    let mut archive = Archive::new(tar);
    archive.set_preserve_permissions(true);
    archive.set_preserve_mtime(true);
    archive.unpack(&dir)?;
    Ok(TestSource::new(dir))
}

// helper function to do windows-specific snapshots (needed e.g. if paths are contained in the snapshot)
fn assert_with_win<T: Serialize>(test: &str, snap: T) {
    #[cfg(windows)]
    assert_ron_snapshot!(format!("{test}-windows"), snap);
    #[cfg(not(windows))]
    assert_ron_snapshot!(format!("{test}-nix"), snap);
}
// Parts of the snapshot summary we want to test against references
//
// # Note
//
// We use a struct to avoid having to escape the field names in the snapshot
// we use insta redactions to replace the actual values with placeholders in case
// there are changes in the actual values
// Readme: https://insta.rs/docs/redactions/
#[derive(Serialize)]
struct TestSummary<'a>(&'a SnapshotFile);

#[rstest]
fn test_backup_with_tar_gz_passes(
    tar_gz_testdata: Result<TestSource>,
    set_up_repo: Result<RepoOpen>,
    insta_summary_redaction: Settings,
    insta_tree_redaction: Settings,
) -> Result<()> {
    // uncomment for logging output
    // SimpleLogger::init(log::LevelFilter::Debug, Config::default())?;

    // Fixtures
    let (source, repo) = (tar_gz_testdata?, set_up_repo?.to_indexed_ids()?);

    let paths = &source.path_list();

    // we use as_path to not depend on the actual tempdir
    let opts = BackupOptions::default().as_path(PathBuf::from_str("test")?);

    // first backup
    let first_snapshot = repo.backup(&opts, paths, SnapshotFile::default())?;

    // We can also bind to scope ( https://docs.rs/insta/latest/insta/struct.Settings.html#method.bind_to_scope )
    // But I think that can get messy with a lot of tests, also checking which settings are currently applied
    // will be probably harder
    insta_summary_redaction.bind(|| {
        assert_with_win("backup-tar-summary-first", TestSummary(&first_snapshot));
    });

    assert_eq!(first_snapshot.parent, None);

    // tree of first backup
    // re-read index
    let repo = repo.to_indexed_ids()?;
    let tree = repo.node_from_path(first_snapshot.tree, Path::new("test/0/tests"))?;
    let tree: rustic_core::repofile::Tree = repo.get_tree(&tree.subtree.expect("Sub tree"))?;

    insta_tree_redaction.bind(|| {
        assert_with_win("backup-tar-tree", tree);
    });

    // get all snapshots and check them
    let all_snapshots = repo.get_all_snapshots()?;
    assert_eq!(vec![first_snapshot.clone()], all_snapshots);
    // save list of pack files
    let packs1: Vec<_> = repo.list(rustic_core::FileType::Pack)?.collect();

    // re-read index
    let repo = repo.to_indexed_ids()?;
    // second backup
    let second_snapshot = repo.backup(&opts, paths, SnapshotFile::default())?;

    insta_summary_redaction.bind(|| {
        assert_with_win("backup-tar-summary-second", TestSummary(&second_snapshot));
    });

    assert_eq!(second_snapshot.parent, Some(first_snapshot.id));
    assert_eq!(first_snapshot.tree, second_snapshot.tree);

    // get all snapshots and check them
    let mut all_snapshots = repo.get_all_snapshots()?;
    all_snapshots.sort_unstable();
    assert_eq!(vec![first_snapshot, second_snapshot], all_snapshots);

    // pack files should be unchanged
    let packs2: Vec<_> = repo.list(rustic_core::FileType::Pack)?.collect();
    assert_eq!(packs1, packs2);

    // Check if snapshots can be retrieved
    let mut ids: Vec<_> = all_snapshots.iter().map(|sn| sn.id.to_string()).collect();
    let snaps = repo.get_snapshots(&ids)?;
    assert_eq!(snaps, all_snapshots);

    // reverse order
    all_snapshots.reverse();
    ids.reverse();
    let snaps = repo.get_snapshots(&ids)?;
    assert_eq!(snaps, all_snapshots);

    Ok(())
}

#[rstest]
fn test_backup_dry_run_with_tar_gz_passes(
    tar_gz_testdata: Result<TestSource>,
    set_up_repo: Result<RepoOpen>,
    insta_summary_redaction: Settings,
    insta_tree_redaction: Settings,
) -> Result<()> {
    // Fixtures
    let (source, repo) = (tar_gz_testdata?, set_up_repo?.to_indexed_ids()?);

    let paths = &source.path_list();

    // we use as_path to not depend on the actual tempdir
    let opts = BackupOptions::default()
        .as_path(PathBuf::from_str("test")?)
        .dry_run(true);

    // dry-run backup
    let snap_dry_run = repo.backup(&opts, paths, SnapshotFile::default())?;

    insta_summary_redaction.bind(|| {
        assert_with_win("dryrun-tar-summary-first", TestSummary(&snap_dry_run));
    });

    // check that repo is still empty
    let snaps = repo.get_all_snapshots()?;
    assert_eq!(snaps.len(), 0);
    assert_eq!(repo.list(rustic_core::FileType::Pack)?.count(), 0);
    assert_eq!(repo.list(rustic_core::FileType::Index)?.count(), 0);

    // first real backup
    let opts = opts.dry_run(false);
    let first_snapshot = repo.backup(&opts, paths, SnapshotFile::default())?;
    assert_eq!(snap_dry_run.tree, first_snapshot.tree);
    let packs: Vec<_> = repo.list(rustic_core::FileType::Pack)?.collect();

    // tree of first backup
    // re-read index
    let repo = repo.to_indexed_ids()?;
    let tree = repo.node_from_path(first_snapshot.tree, Path::new("test/0/tests"))?;
    let tree = repo.get_tree(&tree.subtree.expect("Sub tree"))?;

    insta_tree_redaction.bind(|| {
        assert_with_win("dryrun-tar-tree", tree);
    });

    // re-read index
    let repo = repo.to_indexed_ids()?;
    // second dry-run backup
    let opts = opts.dry_run(true);
    let snap_dry_run = repo.backup(&opts, paths, SnapshotFile::default())?;

    insta_summary_redaction.bind(|| {
        assert_with_win("dryrun-tar-summary-second", TestSummary(&snap_dry_run));
    });

    // check that no data has been added
    let snaps = repo.get_all_snapshots()?;
    assert_eq!(snaps, vec![first_snapshot]);
    let packs_dry_run: Vec<_> = repo.list(rustic_core::FileType::Pack)?.collect();
    assert_eq!(packs_dry_run, packs);

    // re-read index
    let repo = repo.to_indexed_ids()?;
    // second real backup
    let opts = opts.dry_run(false);
    let second_snapshot = repo.backup(&opts, paths, SnapshotFile::default())?;
    assert_eq!(snap_dry_run.tree, second_snapshot.tree);
    Ok(())
}

#[rstest]
fn test_ls(tar_gz_testdata: Result<TestSource>, set_up_repo: Result<RepoOpen>) -> Result<()> {
    // Fixtures
    let (source, repo) = (tar_gz_testdata?, set_up_repo?.to_indexed_ids()?);
    let paths = &source.path_list();

    // we use as_path to not depend on the actual tempdir
    let opts = BackupOptions::default().as_path(PathBuf::from_str("test")?);
    // backup test-data
    let snapshot = repo.backup(&opts, paths, SnapshotFile::default())?;

    // test non-existing entries
    let mut node = Node::new_node(
        OsStr::new(""),
        rustic_core::repofile::NodeType::Dir,
        Metadata::default(),
    );
    node.subtree = Some(snapshot.tree);

    // re-read index
    let repo = repo.to_indexed_ids()?;

    let _entries: Vec<_> = repo
        .ls(&node, &LsOptions::default())?
        .collect::<RusticResult<_>>()?;
    // TODO: Snapshot-test entries
    // assert_ron_snapshot!("ls", entries);
    Ok(())
}

#[rstest]
fn test_find(tar_gz_testdata: Result<TestSource>, set_up_repo: Result<RepoOpen>) -> Result<()> {
    // Fixtures
    let (source, repo) = (tar_gz_testdata?, set_up_repo?.to_indexed_ids()?);
    let paths = &source.path_list();

    // we use as_path to not depend on the actual tempdir
    let opts = BackupOptions::default().as_path(PathBuf::from_str("test")?);
    // backup test-data
    let snapshot = repo.backup(&opts, paths, SnapshotFile::default())?;

    // re-read index
    let repo = repo.to_indexed_ids()?;

    // test non-existing path
    let not_found = repo.find_nodes_from_path(vec![snapshot.tree], Path::new("not_existing"))?;
    assert_with_win("find-nodes-not-found", not_found);
    // test non-existing match
    let glob = Glob::new("not_existing")?.compile_matcher();
    let not_found =
        repo.find_matching_nodes(vec![snapshot.tree], &|path, _| glob.is_match(path))?;
    assert_with_win("find-matching-nodes-not-found", not_found);

    // test existing path
    let FindNode { matches, .. } =
        repo.find_nodes_from_path(vec![snapshot.tree], Path::new("test/0/tests/testfile"))?;
    assert_with_win("find-nodes-existing", matches);
    // test existing match
    let glob = Glob::new("testfile")?.compile_matcher();
    let match_func = |path: &Path, _: &Node| {
        glob.is_match(path) || path.file_name().is_some_and(|f| glob.is_match(f))
    };
    let FindMatches { paths, matches, .. } =
        repo.find_matching_nodes(vec![snapshot.tree], &match_func)?;
    assert_with_win("find-matching-existing", (paths, matches));
    // test existing match
    let glob = Glob::new("testfile*")?.compile_matcher();
    let match_func = |path: &Path, _: &Node| {
        glob.is_match(path) || path.file_name().is_some_and(|f| glob.is_match(f))
    };
    let FindMatches { paths, matches, .. } =
        repo.find_matching_nodes(vec![snapshot.tree], &match_func)?;
    assert_with_win("find-matching-wildcard-existing", (paths, matches));
    Ok(())
}
