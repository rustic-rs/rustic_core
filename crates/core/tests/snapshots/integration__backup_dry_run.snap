---
source: crates/core/tests/integration.rs
expression: TestSnap(&snap_dry_run)
---
TestSnap {
    files_new: 73,
    files_changed: 0,
    files_unmodified: 0,
    total_files_processed: 73,
    total_bytes_processed: 1125674,
    dirs_new: 7,
    dirs_changed: 0,
    dirs_unmodified: 0,
    total_dirs_processed: 7,
    data_blobs: 70,
    tree_blobs: 7,
    data_added_files: 1125653,
    data_added_files_packed: 78740,
}
